include/ginkgo/core/solver/solver_base.hpp

```diff
@@ -50,6 +51,142 @@ namespace gko {
+ */
+class ApplyGuess {
+protected:
+    friend class MultigridState;
+
+    /**
+     * Modify the input vector x by the guess
+     *
+     * @param b  the right hand side vectors
+     * @param x  the input vectors
+     * @param guess  the input guess
+     */
+    static void modify_input(const LinOp* b, LinOp* x, initial_guess guess)
+    {
+        if (guess == initial_guess::zero) {
+            if (auto dense = dynamic_cast<matrix::Dense<float>*>(x)) {
+                dense->fill(zero<float>());
+            } else if (auto dense = dynamic_cast<matrix::Dense<double>*>(x)) {
+                dense->fill(zero<double>());
+            } else if (auto dense =
+                           dynamic_cast<matrix::Dense<std::complex<float>>*>(
+                               x)) {
+                dense->fill(zero<std::complex<float>>());
+            } else if (auto dense =
+                           dynamic_cast<matrix::Dense<std::complex<double>>*>(
+                               x)) {
+                dense->fill(zero<std::complex<double>>());
+            } else {
+                GKO_NOT_IMPLEMENTED;
+            }
+        } else if (guess == initial_guess::rhs) {
+            x->copy_from(b);
+        }
+    }
+
+    /**
+     * Applies a linear operator to a vector (or a sequence of vectors) with
+     * initial guess statement.
+     *
+     * Performs the operation x = op(b) with a initial guess statement, where op
+     * is this linear operator and initial guess statement will change the input
+     * vectors to fit the initial guess.
+     *
+     * @param b  the input vector(s) on which the operator is applied
+     * @param x  the output vector(s) where the result is stored
+     * @param guess  the input guess to handle the input vector(s)
+     *
+     * @return this
+     */
+    virtual void apply_with_guess(const LinOp* b, LinOp* x,
+                                  initial_guess guess) const
+    {
+        modify_input(b, x, guess);
+        this->apply_impl(b, x);
+    }
+
+    /**
+     * Performs the operation x = alpha * op(b) + beta * x with a initial guess
+     * statement, where op is this linear operator and initial guess statement
+     * will change the input vectors to fit the initial guess
+     *
+     * @param alpha  scaling of the result of op(b)
+     * @param b  vector(s) on which the operator is applied
+     * @param beta  scaling of the input x
+     * @param x  output vector(s)
+     * @param guess  the input guess
+     *
+     * @return this
+     */
+    virtual void apply_with_guess(const LinOp* alpha, const LinOp* b,
+                                  const LinOp* beta, LinOp* x,
+                                  initial_guess guess) const
+    {
+        modify_input(b, x, guess);
+        this->apply_impl(alpha, b, beta, x);
+    }
+
+    /**
+     * Get the input guess
+     *
+     * @return initial_guess
+     */
+    initial_guess get_apply_guess() const { return guess_; }
```

Comment:
I think we can be more specific in this naming. Since we only have initial guesses for `apply`, we don't need to prefix it here.
```suggestion
    initial_guess get_default_initial_guess() const { return guess_; }
```